File: frontend/src/components/utils/userTags.js
export const USER_TAG_GROUPS = [
  {
    id: "sound-traits",
    label: "Sound Traits",
    tags: [
      { id: "atmospheric", label: "Atmospheric" },
      { id: "punchy", label: "Punchy" },
      { id: "lush", label: "Lush" },
      { id: "raw", label: "Raw" },
      { id: "hazy", label: "Hazy" },
      { id: "sleek", label: "Sleek" },
      { id: "layered", label: "Layered" },
      { id: "groove-heavy", label: "Groove-heavy" },
      { id: "sparse", label: "Sparse" },
      { id: "noisy", label: "Noisy" },
    ],
  },
  {
    id: "emotional-tone",
    label: "Emotional Tone",
    tags: [
      { id: "melancholic", label: "Melancholic" },
      { id: "nostalgic", label: "Nostalgic" },
      { id: "euphoric", label: "Euphoric" },
      { id: "tense", label: "Tense" },
      { id: "reflective", label: "Reflective" },
      { id: "bittersweet", label: "Bittersweet" },
      { id: "playful", label: "Playful" },
      { id: "romantic", label: "Romantic" },
      { id: "defiant", label: "Defiant" },
      { id: "lonely", label: "Lonely" },
      { id: "hopeful", label: "Hopeful" },
    ],
  },
  {
    id: "album-qualities",
    label: "Album Qualities",
    tags: [
      { id: "cohesive", label: "Cohesive" },
      { id: "addictive", label: "Addictive" },
      { id: "ambitious", label: "Ambitious" },
      { id: "polished", label: "Polished" },
      { id: "experimental", label: "Experimental" },
      { id: "accessible", label: "Accessible" },
      { id: "rewarding", label: "Rewarding" },
      { id: "timeless", label: "Timeless" },
      { id: "uneven", label: "Uneven" },
      { id: "distinctive", label: "Distinctive" },
    ],
  },
];

const USER_TAG_LABELS = Object.fromEntries(
  USER_TAG_GROUPS.flatMap((group) => group.tags.map((tag) => [tag.id, tag.label]))
);

export function getUserTagLabel(tagId) {
  return USER_TAG_LABELS[tagId] || tagId;
}

File: backend/app/user_tags.py
from collections.abc import Iterable


USER_TAG_GROUPS: dict[str, tuple[tuple[str, str], ...]] = {
    "sound_traits": (
        ("atmospheric", "Atmospheric"),
        ("punchy", "Punchy"),
        ("lush", "Lush"),
        ("raw", "Raw"),
        ("hazy", "Hazy"),
        ("sleek", "Sleek"),
        ("layered", "Layered"),
        ("groove-heavy", "Groove-heavy"),
        ("sparse", "Sparse"),
        ("noisy", "Noisy"),
    ),
    "emotional_tone": (
        ("melancholic", "Melancholic"),
        ("nostalgic", "Nostalgic"),
        ("euphoric", "Euphoric"),
        ("tense", "Tense"),
        ("reflective", "Reflective"),
        ("bittersweet", "Bittersweet"),
        ("playful", "Playful"),
        ("romantic", "Romantic"),
        ("defiant", "Defiant"),
        ("lonely", "Lonely"),
        ("hopeful", "Hopeful"),
    ),
    "album_qualities": (
        ("cohesive", "Cohesive"),
        ("addictive", "Addictive"),
        ("ambitious", "Ambitious"),
        ("polished", "Polished"),
        ("experimental", "Experimental"),
        ("accessible", "Accessible"),
        ("rewarding", "Rewarding"),
        ("timeless", "Timeless"),
        ("uneven", "Uneven"),
        ("distinctive", "Distinctive"),
    ),
}

ALLOWED_USER_TAGS = {
    tag_id
    for tags in USER_TAG_GROUPS.values()
    for tag_id, _label in tags
}


def normalize_user_tags(values: Iterable[str] | None) -> list[str]:
    normalized: list[str] = []
    seen: set[str] = set()

    for raw_value in values or []:
        value = str(raw_value or "").strip()
        if not value or value in seen:
            continue
        if value not in ALLOWED_USER_TAGS:
            raise ValueError(f"Unsupported user tag: {value}")
        seen.add(value)
        normalized.append(value)

    return normalized
